src/phongo_apm.c

```diff
@@ -0,0 +1,297 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "bson/bson.h"
+#include "mongoc/mongoc.h"
+
+#include <php.h>
+#include <Zend/zend_hash.h>
+#include <Zend/zend_exceptions.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+#include "phongo_apm.h"
+
+#undef MONGOC_LOG_DOMAIN
+#define MONGOC_LOG_DOMAIN "PHONGO"
+
+ZEND_EXTERN_MODULE_GLOBALS(mongodb)
+
+/* Ensures that all subscribers in @from that are instances of @subscriber_ce
+ * are added to the set of subscribers in @to. This is used to build the list of
+ * subscribers to notify for an event. */
+static void phongo_apm_add_subscribers(zend_class_entry* subscriber_ce, HashTable* from, HashTable* to)
+{
+	zval* subscriber;
+
+	ZEND_HASH_FOREACH_VAL_IND(from, subscriber)
+	{
+		if (Z_TYPE_P(subscriber) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(subscriber), subscriber_ce)) {
+			continue;
+		}
+
+		if (zend_hash_index_exists(to, Z_OBJ_HANDLE_P(subscriber))) {
+			continue;
+		}
+
+		zend_hash_index_update(to, Z_OBJ_HANDLE_P(subscriber), subscriber);
+		Z_ADDREF_P(subscriber);
+	}
+	ZEND_HASH_FOREACH_END();
+}
+
+/* Returns a newly allocated HashTable, which will contain all subscribers of a
+ * certain type that should be notified for an event on the specified client. */
+static HashTable* phongo_apm_get_subscribers(zend_class_entry* subscriber_ce, mongoc_client_t* client)
+{
+	HashTable* subscribers = NULL;
+
+	ALLOC_HASHTABLE(subscribers);
+	zend_hash_init(subscribers, 0, NULL, ZVAL_PTR_DTOR, 0);
+
+	if (MONGODB_G(subscribers)) {
+		phongo_apm_add_subscribers(subscriber_ce, MONGODB_G(subscribers), subscribers);
+	}
+
+	if (MONGODB_G(managers)) {
+		php_phongo_manager_t* manager;
+
+		ZEND_HASH_FOREACH_PTR(MONGODB_G(managers), manager)
+		{
+			if (manager->client == client && manager->subscribers) {
+				phongo_apm_add_subscribers(subscriber_ce, manager->subscribers, subscribers);
+			}
+		}
+		ZEND_HASH_FOREACH_END();
+	}
+
+	return subscribers;
+}
+
+/* Search for a Manager associated with the given client in the request-scoped
+ * registry. If any Manager is found, copy it to @out, increment its ref-count,
+ * and return true; otherwise, set @out to undefined and return false. */
+static bool phongo_apm_copy_manager_for_client(mongoc_client_t* client, zval* out)
+{
+	php_phongo_manager_t* manager;
+
+	ZVAL_UNDEF(out);
```

Comment:
While moving this function from `php_phongo.c` to here, I moved this line up. Previously, it was only called for the final `return false` and never for the `return false` in the condition that immediately follows. That said, I don't think this changes anything I believe `php_phongo_commandstartedevent_t.manager` will be initialized to zero anyway.